homeassistant/components/google_photos/media_source.py

```diff
@@ -0,0 +1,291 @@
+"""Media source for Google Photos."""
+
+from dataclasses import dataclass
+import logging
+from typing import Any, cast
+
+from homeassistant.components.media_player import MediaClass, MediaType
+from homeassistant.components.media_source import (
+    BrowseError,
+    BrowseMediaSource,
+    MediaSource,
+    MediaSourceItem,
+    PlayMedia,
+)
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import HomeAssistant
+
+from . import GooglePhotosConfigEntry
+from .const import DOMAIN
+from .exceptions import GooglePhotosApiError
+
+_LOGGER = logging.getLogger(__name__)
+
+# Media Sources do not support paging, so we only show a subset of recent
+# photos when displaying the users library. We fetch a minimum of 50 photos
+# unless we run out, but in pages of 100 at a time given sometimes responses
+# may only contain a handful of items Fetches at least 50 photos.
+MAX_PHOTOS = 50
+PAGE_SIZE = 100
+
+THUMBNAIL_SIZE = 256
+LARGE_IMAGE_SIZE = 2048
+
+
+# Markers for parts of PhotosIdentifier url pattern.
+# The PhotosIdentifier can be in the following forms:
+#  config-entry-id
+#  config-entry-id/a/album-media-id
+#  config-entry-id/p/photo-media-id
+#
+# The album-media-id can contain special reserved folder names for use by
+# this integration for virtual folders like the `recent` album.
+PHOTO_SOURCE_IDENTIFIER_PHOTO = "p"
+PHOTO_SOURCE_IDENTIFIER_ALBUM = "a"
+
+# Currently supports a single album of recent photos
+RECENT_PHOTOS_ALBUM = "recent"
+RECENT_PHOTOS_TITLE = "Recent Photos"
```

Comment:
I can see how to do a one-off for google photos translations, however that doesn't appear to be a common pattern. I think if we want to support this it would make sense to do for all media sources. I spent about 45 minutes looking at what it might take to do this and am pausing as it looks quite involved given the current media player APIs. If there is a simple common pattern here i'm missing happy follow it.

Reply:
I just looked at it too. It doesn't seem it's possible without a hack such as adding the string as an exception. Maybe introduce a generic strings section in strings.json and add a function similar to async_get_exception_message in helpers.translation? The functions could share most of the code. Do you think you could open a design discussion?